opentelemetry-auto-instrumentation/src/opentelemetry/auto_instrumentation/bootstrap.py

```diff
@@ -0,0 +1,177 @@
+    "pymongo": "opentelemetry-ext-pymongo>=0.6b0",
+    "pymysql": "opentelemetry-ext-pymysql",
+    "redis": "opentelemetry-ext-redis",
+    "sqlalchemy": "opentelemetry-ext-sqlalchemy",
+    "wsgi": "opentelemetry-ext-wsgi>=0.6b0",
+}
+
+# relevant instrumentors and tracers to uninstall and check for conflicts for target libraries
+libraries = {
+    "dbapi": ("opentelemetry-ext-dbapi",),
+    "flask": ("opentelemetry-ext-flask",),
+    "grpc": ("opentelemetry-ext-grpc",),
+    "requests": ("opentelemetry-ext-requests",),
+    "mysql": ("opentelemetry-ext-mysql",),
+    "psycopg2": ("opentelemetry-ext-psycopg2",),
+    "pymongo": ("opentelemetry-ext-pymongo",),
+    "pymysql": ("opentelemetry-ext-pymysql",),
+    "redis": ("opentelemetry-ext-redis",),
+    "sqlalchemy": ("opentelemetry-ext-sqlalchemy",),
+    "wsgi": ("opentelemetry-ext-wsgi",),
+}
+
+
+def _install_package(library, instrumentation):
+    """
+    Ensures that desired version is installed w/o upgrading its dependencies
+    by uninstalling where necessary (if `target` is not provided).
+
+
+    OpenTelemetry auto-instrumentation packages often have traced libraries
+    as instrumentation dependency (e.g. flask for opentelemetry-ext-flask),
+    so using -I on library could cause likely undesired Flask upgrade.
+    Using --no-dependencies alone would leave potential for nonfunctional
+    installations.
+    """
+    pip_list = _pip_freeze()
+    for package in libraries[library]:
+        if "{}==".format(package).lower() in pip_list:
+            logger.info(
+                "Existing %s installation detected.  Uninstalling.", package
+            )
+            _pip_uninstall(package)
+    _pip_install(instrumentation)
+
+
+def _pip_freeze():
+    return (
+        subprocess.check_output([sys.executable, "-m", "pip", "freeze"])
+        .decode()
+        .lower()
+    )
+
+
+def _pip_install(package):
+    # explicit upgrade strategy to override potential pip config
+    subprocess.check_call(
+        [
+            sys.executable,
+            "-m",
+            "pip",
+            "install",
+            "-U",
+            "--upgrade-strategy",
+            "only-if-needed",
+            package,
+        ]
+    )
+
+
+def _pip_uninstall(package):
+    subprocess.check_call(
+        [sys.executable, "-m", "pip", "uninstall", "-y", package]
+    )
+
+
+def _pip_check():
+    """Ensures none of the instrumentations have dependency conflicts.
+    Clean check reported as:
+    'No broken requirements found.'
+    Dependency conflicts are reported as:
+    'opentelemetry-ext-flask 1.0.1 has requirement opentelemetry-sdk<2.0,>=1.0, but you have opentelemetry-sdk 0.5.'
+    To not be too restrictive, we'll only check for relevant packages.
+    """
+    check_pipe = subprocess.Popen(
+        [sys.executable, "-m", "pip", "check"], stdout=subprocess.PIPE
+    )
+    pip_check = check_pipe.communicate()[0].decode()
+    pip_check_lower = pip_check.lower()
```

Comment:
Curious as to why we can't simply string search for "opentelemetry-ext" here?

Reply:
I think it's better to be more explicit, check for each library and to not encode the knowledge about libraries in multiple places. Right now the `libraries` mapping remains the single source of truth for all libraries we support. Adding sub-string matching on common prefix leaks that knowledge into this function. I don't much benefit in doing that. Rest of the program follows "configuration over convention" and I'd prefer to continue doing that but happy to hear other thoughts.

Reply:
That's a good reason. Would there be the case that some other library has a dependency conflict (not an integration) with one of the dependencies and then will generate a false positive?

Reply:
@owais

Reply:
It is possible I suppose but not sure if it we should try to account for all possible conflicts globally in site-packages. Such conflicts could also be present independently of the bootstrap command so not sure if bootstrap should complain about them. In such case, users can use the default option to print out a list of dependencies instead of installing them and then figure out the conflicts as they would with any other library. Happy to hear other thoughts.